### ISSUE TITLE
Crash in mixpanel-iphone v3.3.7

### ISSUE THREAD

**Issue by 0bj** [mixpanel/mixpanel-iphone#816]
**Integration Method:** CocoaPods
**Xcode Version:** 10.0
**Library Version:** 3.3.7
**Platform:** iOS
**Language:** Swift & Objective-C - mixed
**Description:** I got this crash on iPhone 5C(10.3.3), seems to be related to the "identify:usePeople:" method call ([Mixpanel identify:usePeople:]_block_invoke (Mixpanel.m:454)). App installed from TestFlight (Release mode). This crash does not happen consistently.
**Expected Behavior:** 
The SDK shouldn't crash.

![image](https://user-images.githubusercontent.com/12246254/48599734-d601c900-e98a-11e8-8624-47073d1e3a24.png)

454 line:
![screenshot 2018-11-16 at 10 43 34](https://user-images.githubusercontent.com/12246254/48600144-8c19e280-e98c-11e8-9677-125b802290ea.png)

**Comment by zihejia** [comment]
Hi @0bj Sergey, thanks for raising it. Is your code calling the `identify:usePeople:` written in swift?

**Comment by 0bj** [comment]
Hi @zihejia, objective-c

**Comment by zioyero** [comment]
I'm also experiencing a crash seemingly related to using this backing array. Whenever the app is put into the background, I get a `-[__NSArrayI addObject:] unrecognized selector sent to instance` error.

I can create a new issue if you believe this is a separate problem.

**Comment by zioyero** [comment]
We were able to narrow down how to explicitly reproduce this crash. It happens when you downgrade from 3.3.7 to a previous version, then update again. 

Exact steps:

1. Clean installation of an app with Mixpanel SDK version 3.3.7
2. Update app with Mixpanel SDK version 3.3.3
3. Update app with Mixpanel SDK version 3.3.7
4. Put the app in the background - Crash

For our particular use case, being able to downgrade to an older version of the app and then upgrade is important, but not critical, so the simple workaround is to delete and reinstall the app with the updated version.

Hope this helps.

**Comment by abotkin-cpi** [comment]
I've also hit this.  It appears to be happening due to the unidentifiedQueue having a non-mutable NSArray set.  Seems to be due to the properties archive having the properties[@"peopleUnidentifiedQueue"] return an immutable array, which is directly set onto a mutable array property without any mutableCopy in Mixpanel.m:1152

```
(lldb) po self.unidentifiedQueue
<__NSArrayI 0x60000118fc60>(
{
    "$mp_metadata" =     {
        "$mp_event_id" = c0ebc3455a0e368d;
        "$mp_session_id" = 06056a37397c5ae8;
        "$mp_session_seq_id" = 0;
        "$mp_session_start_sec" = 1541471649;
    };
    "$set_once" =     {
        "$ae_first_app_open_date" = "2018-11-06 02:34:09 +0000";
        "$ios_app_release" = 142;
        "$ios_app_version" = "5.38.1";
        "$ios_device_model" = "x86_64";
        "$ios_ifa" = "<redacted>";
        "$ios_lib_version" = "3.3.6";
        "$ios_version" = "12.1";
    };
    "$time" = 1541471649119;
    "$token" = <redacted>;
},
{
    "$mp_metadata" =     {
        "$mp_event_id" = 29f914a8354def80;
        "$mp_session_id" = 06056a37397c5ae8;
        "$mp_session_seq_id" = 1;
        "$mp_session_start_sec" = 1541471649;
    };
    "$set" =     {
        "$ios_app_release" = 142;
        "$ios_app_version" = "5.38.1";
        "$ios_device_model" = "x86_64";
        "$ios_ifa" = "<redacted>";
        "$ios_lib_version" = "3.3.6";
        "$ios_version" = "12.1";
        "<redacted>" = 0;
    };
    "$time" = 1541471651472;
    "$token" = <redacted>;
}
)

(lldb) po [self.unidentifiedQueue isKindOfClass:[NSMutableArray class]]
NO

(lldb) po [self.unidentifiedQueue isKindOfClass:[NSArray class]]
YES
```

**Comment by zihejia** [comment]
To both @abotkin-cpi and @zioyero, thanks so much for all your detail observations and you are correct! It should be good moving forward, but for downgrade scenario, the workaround is to apply the commit  56e51e6 as a patch manually to any pre v3.3.7 version of Mixpanel. 

In terms of the shallow copy, it's a trade off, since it's very cheap and most crash was caused by the first level of mutation. But we will keep an eye on this, if the modification still happens from time to time, we will continue to work on it.

**Comment by zihejia** [comment]
hi @gilgol , you may open a new issue for 3.3.7 and share the crash log if it's not the same one. Thanks for pointing it out.

**Comment by MayeuldP** [comment]
> We were able to narrow down how to explicitly reproduce this crash. It happens when you downgrade from 3.3.7 to a previous version, then update again.
> 
> Exact steps:
> 
> 1. Clean installation of an app with Mixpanel SDK version 3.3.7
> 2. Update app with Mixpanel SDK version 3.3.3
> 3. Update app with Mixpanel SDK version 3.3.7
> 4. Put the app in the background - Crash
> 
> For our particular use case, being able to downgrade to an older version of the app and then upgrade is important, but not critical, so the simple workaround is to delete and reinstall the app with the updated version.
> 
> Hope this helps.

I also had the same error message when putting the app in background, I downgrade to `3.3.3` but I still get the error. And I don't understand the workaround, because if I do an `archive` through Xcode, wether it've been installed on the simulator before or not, it's gonna recompiled everything, no, so doesn't care about the current version on the simulator ? :/

**Comment by zihejia** [comment]
To answer @MayeuldP 's question, reinstall on the simulator or device won't clear the cache files unfortunately. 
We have released v3.3.9 which has a fix for the crashes caused by v3.3.7. Again, sorry about the inconvenience and please feel free to let me know if the problem still exists and I will reopen it.
